### PR TITLE
(#4449) - add custom builds

### PR DESCRIPTION
**Pull request by nolanlawson**
Adds the ability to customize the PouchDB build, using either Browserify/Webpack or by just downloading from the website.

Adds some complexity to our build process (see changes to `CONTRIBUTING.md`) at the expense of an arguably very very useful feature for our users.

There is a new `npm run build-custom` that builds up all possible combinations of builds, calculates the min/gzip size, and moves them to the `static/js` directory. These builds are not published to npm/bower; I think they're too big for that (~12MB).

Not implemented yet: separating out migrations, separating out `lie`.

animated

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -1,23 +1,10 @@
 "use strict";
```
Comment (nolanlawson) [inline lib/index.js]: Note that PouchDB itself is now a custom build. It's important to eat your own dog food. :smiley:

**Comment by nolanlawson** [comment]
Also `require('pouchdb/custom'leveldb')` is in kind of a weird position, because I can't imagine why any npm user would want to use it, but it's in there for our own custom build. My current solution is to just not document it, because I don't think it's necessary.

**Comment by nolanlawson** [comment]
I should also do a blog post explaining what you might use this for:

* Using only in-memory, WorkerPouch, or some other adapter? Skip both the IndexedDB and the WebSQL adapters.
* Want to use PouchDB just as an interface to CouchDB? Uncheck everything except "HTTP."
* Want to skip map/reduce? Go ahead; you'll save 4KB.

**Comment by nickcolley** [comment]
This is really cool, can you give me some time to review (hopefully later today)

Thinking it'd make more sense to have this in the downloads page?

**Comment by nolanlawson** [comment]
@nickcolley Yeah, I wasn't really sure where to put it. Also would definitely like your help with the design/CSS/etc. :)

**Comment by nolanlawson** [comment]
Rebased, also fixed it so that `ajax` is not included if you don't choose "HTTP". Makes that build a bit smaller.

**Comment by nickcolley** [comment]
Started making some changes to make this a downloads page.

One thing I want to add is the ability to set the build via params in the URL to make it easier to share a custom build.

**Comment by nolanlawson** [comment]
> One thing I want to add is the ability to set the build via params in the URL to make it easier to share a custom build.

I worry that's potentially dangerous, because then people might start using `pouchdb.com` as a CDN, which is a problem we ran into with `pouchb.com/pouchdb-nightly.js`. I kind of intended to make the url non-human-readable (using the base64 hash), so that people wouldn't feel encouraged to use them in that way. :/

**Comment by nickcolley** [comment]
I meant the URL to the page, not the physical build itself. Agree with what you're saying.

e.g. `pouchdb.com/downloads?custom=http` or whatever. So we can link to the download page.

**Comment by nolanlawson** [comment]
Ah okay, so it would pre-check the boxes depending on the query params? Sounds like a nifty feature.

**Comment by nolanlawson** [comment]
I think I would like to put custom builds on the back burner for the time being. From hanging out with @rich-harris and @reconbot and talking to them about Rollup/ES6

**Pull request closed without merging**